doc/aioctl.8
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "AIOCTL" "8" "November 2019" "Datto Inc" "aioctl Manual"
.
.SH "NAME"
\fBaioctl\fR
.
.SH "NAME"
assurio-snap \- Control the Assurio-Snap block device kernel module\.
.
.SH "SYNOPSIS"
\fBaioctl <sub\-command> [<args>]\fR
.
.SH "DESCRIPTION"
\fBaioctl\fR is the userspace tool used to manage the assurio-snap kernel module\. It provides an interface to create, delete, reload, transition, and configure on\-disk snapshots and certain parameters of the kernel module itself\.
.
.P
This manual page describes \fBaioctl\fR briefly\. More detail is available in the Git repository located at https://github\.com/assuriosw/assurio-snap\.
.
.SH "OPTIONS"
.
.nf

\-c cache\-size
     Specify how big the in\-memory data cache can grow to (in MB)\. Defaults to 300 MB\.

\-f fallocate
     Specify the maximum size of the COW file on disk\.
.
.fi
.
.SH "SUB\-COMMANDS"
.
.SS "setup\-snapshot"
\fBaioctl setup\-snapshot [\-c <cache size>] [\-f <fallocate>] <block device> <cow file path> <minor>\fR
.
.P
Sets up a snapshot of \fB<block device>\fR, saving all COW data to \fB<cow file path>\fR\. The snapshot device will be \fB/dev/assurio-snap<minor>\fR\. The minor number will be used as a reference number for all other \fBaioctl\fR commands\. \fB<cow file path>\fR must be a path on the \fB<block device>\fR\.
.
.SS "reload\-snapshot"
\fBaioctl reload\-snapshot [\-c <cache size>] <block device> <cow file> <minor>\fR
.
.P
Reloads a snapshot\. This command is meant to be run before the block device is mounted, after a reboot or after the driver is unloaded\. It notifies the kernel driver to expect the block device specified to come back online\. This command requires that the snapshot was cleanly unmounted in snapshot mode beforehand\. If this is not the case, the snapshot will be put into the failure state once it attempts to come online\. The minor number will be used as a reference number for all other \fBaioctl\fR commands\.
.
.SS "reload\-incremental"
\fBaioctl reload\-incremental [\-c <cache size>] <block device> <cow file> <minor>\fR
.
.P
Reloads a block device that was in incremental mode\. See \fBreload\-snapshot\fR for restrictions\.
.
.SS "transition\-to\-incremental"
\fBaioctl transition\-to\-incremental <minor>\fR
.
.P
Transitions a snapshot COW file to incremental mode, which only tracks which blocks have changed since a snapshot started\. This will remove the associated snapshot device\.
.
.SS "transition\-to\-snapshot"
\fBaioctl transition\-to\-snapshot [\-f <fallocate>] <cow file> <minor>\fR
.
.P
Transitions a block device in incremental mode to snapshot mode\. This call ensures no writes are missed between tearing down the incremental and setting up the new snapshot\. The new snapshot data will be recorded in \fB<cow file>\fR\. The old cow file will still exist after this and can be used to efficiently copy only changed blocks using a tool succh as \fBupdate\-img\fR\.
.
.SS "destroy"
\fBaioctl destroy <minor>\fR
.
.P
Cleanly and completely removes the snapshot or incremental, unlinking the associated COW file\.
.
.SS "reconfigure"
\fBaioctl reconfigure [\-c <cache size>] <minor>\fR
.
.P
Allows you to reconfigure various parameters of a snapshot while it is online\. Currently only the index cache size (given in MB) can be changed dynamically\.
.
.SS "EXAMPLES"
\fB# aioctl setup\-snapshot /dev/sda1 /var/backup/assurio 4\fR
.
.P
This command will set up a new COW snapshot device tracking \fB/dev/sda1\fR at \fB/dev/assurio-snap4\fR\. This block device is backed by a new file created at the path \fB/var/backup/assurio\fR\.
.
.P
\fB# aioctl transition\-to\-incremental 4\fR
.
.P
Transitions the snapshot specified by the minor number to incremental mode\.
.
.P
\fB# aioctl transition\-to\-snapshot /var/backup/assurio1 4\fR
.
.P
Cleanly transitions the incremental to a new snapshot, using \fB/var/backup/assurio1\fR as the new COW file\. At this point a second backup can be taken, either doing a full copy with a tool like \fBdd\fR or an incremental copy using a tool such as \fBupdate\-img\fR, if a previous snapshot backup exists\.
.
.P
\fB# aioctl reconfigure \-c 400 4\fR
.
.P
Reconfigures the block device to have an in\-memory index cache size of 400 MB\.
.
.P
\fB# aioctl destroy 4\fR
.
.P
This will stop tracking \fB/dev/sda1\fR, remove the associated \fB/dev/assurio-snap4\fR (since the device is in snapshot mode), delete the COW file backing it, and perform all other cleanup\.
.
.P
\fB# aioctl reload\-snapshot /dev/sda1 /var/backup/assurio1 4\fR
.
.P
After a reboot, this command may be performed in the early stages of boot, before the block device is mounted read\-write\. This will notify the driver to expect a block device \fB/dev/sda1\fR that was left in snapshot mode to come online with a COW file located at \fB/var/backup/assurio1\fR (relative to the mountpoint), and that the reloaded snapshot should come online at minor number 4\. If a problem is discovered when the block device comes online, this block device will be put into the failure state, which will be reported in \fB/proc/assurio\-snap\-info\fR
.
.P
\fB# aioctl reload\-incremental /dev/sda5 /var/backup/assurio1 4\fR
.
.P
This will act the same as \fBreload\-snapshot\fR, but for a device that was left in incremental mode\.
.
.SH "Bugs"
.
.SH "Author"
.
.nf

Tom Caputi (tcaputi@datto\.com)
.
.fi



app/aioctl.c
// SPDX-License-Identifier: GPL-2.0-only

/*
 * Copyright (C) 2015 Datto Inc.
 * Additional contributions by Assurio Software, Inc are Copyright (C) 2019 Assurio Software Inc.
 */

#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <stdlib.h>
#include <errno.h>
#include <limits.h>
#include <ctype.h>

#include "libassurio-snap.h"

static void print_help(int status){
	printf("Usage:\n");
	printf("\taioctl setup-snapshot [-c <cache size>] [-f fallocate] <block device> <cow file> <minor>\n");
	printf("\taioctl reload-snapshot [-c <cache size>] <block device> <cow file> <minor>\n");
	printf("\taioctl reload-incremental [-c <cache size>] <block device> <cow file> <minor>\n");
	printf("\taioctl destroy <minor>\n");
	printf("\taioctl transition-to-incremental <minor>\n");
	printf("\taioctl transition-to-snapshot [-f fallocate] <cow file> <minor>\n");
	printf("\taioctl reconfigure [-c <cache size>] <minor>\n");
	printf("\taioctl help\n\n");
	printf("for the reload commands, <cow file> should be specified relative to the root of <block device>\n");
	printf("cache size should be provided in bytes, and fallocate should be provided in megabytes\n");
	printf("note: if the -c or -f options are not specified for any given call, module defaults are used\n");
	exit(status);
}

static int parse_ul(const char *str, unsigned long *out){
	long long tmp;
	const char *c = str;

	//check that string is an integer number and has a length
	do{
		if(!isdigit(*c)) goto error;
		c++;
	}while(*c);

	//convert to long long
	tmp = strtoll(str, NULL, 0);
	if(errno) goto error;

	//check boundaries
	if(tmp < 0 || tmp == LLONG_MAX){
		errno = ERANGE;
		goto error;
	}

	*out = (unsigned long)tmp;
	return 0;

error:
	*out = 0;
	return -1;
}

static int parse_ui(const char *str, unsigned int *out){
	long tmp;
	const char *c = str;

	//check that string is an integer number and has a length
	do{
		if(!isdigit(*c)) goto error;
		c++;
	}while(*c);

	//convert to long
	tmp = strtol(str, NULL, 0);
	if(errno) goto error;

	//check boundaries
	if(tmp < 0 || tmp == LONG_MAX){
		errno = ERANGE;
		goto error;
	}

	*out = (unsigned int)tmp;
	return 0;

error:
	*out = 0;
	return -1;
}

static int handle_setup_snap(int argc, char **argv){
	int ret, c;
	unsigned int minor;
	unsigned long cache_size = 0, fallocated_space = 0;
	char *bdev, *cow;

	//get cache size and fallocated space params, if given
	while((c = getopt(argc, argv, "c:f:")) != -1){
		switch(c){
		case 'c':
			ret = parse_ul(optarg, &cache_size);
			if(ret) goto error;
			break;
		case 'f':
			ret = parse_ul(optarg, &fallocated_space);
			if(ret) goto error;
			break;
		default:
			errno = EINVAL;
			goto error;
		}
	}

	if(argc - optind != 3){
		errno = EINVAL;
		goto error;
	}

	bdev = argv[optind];
	cow = argv[optind + 1];

	ret = parse_ui(argv[optind + 2], &minor);
	if(ret) goto error;

	return assurio_snap_setup_snapshot(minor, bdev, cow, fallocated_space, cache_size);

error:
	perror("error interpreting setup snapshot parameters");
	print_help(-1);
	return 0;
}

static int handle_reload_snap(int argc, char **argv){
	int ret, c;
	unsigned int minor;
	unsigned long cache_size = 0;
	char *bdev, *cow;

	//get cache size and fallocated space params, if given
	while((c = getopt(argc, argv, "c:")) != -1){
		switch(c){
		case 'c':
			ret = parse_ul(optarg, &cache_size);
			if(ret) goto error;
			break;
		default:
			errno = EINVAL;
			goto error;
		}
	}

	if(argc - optind != 3){
		errno = EINVAL;
		goto error;
	}

	bdev = argv[optind];
	cow = argv[optind + 1];

	ret = parse_ui(argv[optind + 2], &minor);
	if(ret) goto error;

	return assurio_snap_reload_snapshot(minor, bdev, cow, cache_size);

error:
	perror("error interpreting reload snapshot parameters");
	print_help(-1);
	return 0;
}

static int handle_reload_inc(int argc, char **argv){
	int ret, c;
	unsigned int minor;
	unsigned long cache_size = 0;
	char *bdev, *cow;

	//get cache size and fallocated space params, if given
	while((c = getopt(argc, argv, "c:")) != -1){
		switch(c){
		case 'c':
			ret = parse_ul(optarg, &cache_size);
			if(ret) goto error;
			break;
		default:
			errno = EINVAL;
			goto error;
		}
	}

	if(argc - optind != 3){
		errno = EINVAL;
		goto error;
	}

	bdev = argv[optind];
	cow = argv[optind + 1];

	ret = parse_ui(argv[optind + 2], &minor);
	if(ret) goto error;

	return assurio_snap_reload_incremental(minor, bdev, cow, cache_size);

error:
	perror("error interpreting reload incremental parameters");
	print_help(-1);
	return 0;
}

static int handle_destroy(int argc, char **argv){
	int ret;
	unsigned int minor;

	if(argc != 2){
		errno = EINVAL;
		goto error;
	}

	ret = parse_ui(argv[1], &minor);
	if(ret) goto error;

	return assurio_snap_destroy(minor);

error:
	perror("error interpreting destroy parameters");
	print_help(-1);
	return 0;
}

static int handle_transition_inc(int argc, char **argv){
	int ret;
	unsigned int minor;

	if(argc != 2){
		errno = EINVAL;
		goto error;
	}

	ret = parse_ui(argv[1], &minor);
	if(ret) goto error;

	return assurio_snap_transition_incremental(minor);

error:
	perror("error interpreting transition to incremental parameters");
	print_help(-1);
	return 0;
}

static int handle_transition_snap(int argc, char **argv){
	int ret, c;
	unsigned int minor;
	unsigned long fallocated_space = 0;
	char *cow;

	//get cache size and fallocated space params, if given
	while((c = getopt(argc, argv, "f:")) != -1){
		switch(c){
		case 'f':
			ret = parse_ul(optarg, &fallocated_space);
			if(ret) goto error;
			break;
		default:
			errno = EINVAL;
			goto error;
		}
	}

	if(argc - optind != 2){
		errno = EINVAL;
		goto error;
	}

	cow = argv[optind];

	ret = parse_ui(argv[optind + 1], &minor);
	if(ret) goto error;

	return assurio_snap_transition_snapshot(minor, cow, fallocated_space);

error:
	perror("error interpreting transition to snapshot parameters");
	print_help(-1);
	return 0;
}

static int handle_reconfigure(int argc, char **argv){
	int ret, c;
	unsigned int minor;
	unsigned long cache_size = 0;

	//get cache size and fallocated space params, if given
	while((c = getopt(argc, argv, "c:")) != -1){
		switch(c){
		case 'c':
			ret = parse_ul(optarg, &cache_size);
			if(ret) goto error;
			break;
		default:
			errno = EINVAL;
			goto error;
		}
	}

	if(argc - optind != 1){
		errno = EINVAL;
		goto error;
	}

	ret = parse_ui(argv[optind], &minor);
	if(ret) goto error;

	return assurio_snap_reconfigure(minor, cache_size);

error:
	perror("error interpreting reconfigure parameters");
	print_help(-1);
	return 0;
}

int main(int argc, char **argv){
	int ret = 0;

	//check argc
	if(argc < 2) print_help(-1);

	if(access("/dev/assurio-snap-ctl", F_OK) != 0){
		errno = EINVAL;
		perror("driver does not appear to be loaded");
		return -1;
	}

	//route to appropriate handler or print help
	if(!strcmp(argv[1], "setup-snapshot")) ret = handle_setup_snap(argc - 1, argv + 1);
	else if(!strcmp(argv[1], "reload-snapshot")) ret = handle_reload_snap(argc - 1, argv + 1);
	else if(!strcmp(argv[1], "reload-incremental")) ret = handle_reload_inc(argc - 1, argv + 1);
	else if(!strcmp(argv[1], "destroy")) ret = handle_destroy(argc - 1, argv + 1);
	else if(!strcmp(argv[1], "transition-to-incremental")) ret = handle_transition_inc(argc - 1, argv + 1);
	else if(!strcmp(argv[1], "transition-to-snapshot")) ret = handle_transition_snap(argc - 1, argv + 1);
	else if(!strcmp(argv[1], "reconfigure")) ret = handle_reconfigure(argc - 1, argv + 1);
	else if(!strcmp(argv[1], "help")) print_help(0);
	else print_help(-1);

	if(ret) perror("driver returned an error performing specified action. check dmesg for more info");

	return ret;
}
